File: BetESS_a69303_pg25291 11nov_observer/BetESS/src/betess/Apostador.java
package betess;

import java.util.ArrayList;
import java.util.TreeMap;
import java.util.TreeSet;

/**
 *
 * @author ana almeida
 */
public class Apostador extends User implements Observer {

    private double valor;
    private int contadorIdAposta;
    //private HashMap<LocalDateTime, Evento> eventos;
    private TreeMap<Integer, Aposta> listaApostas;
    private TreeSet<Integer> notificacoesOdd; // id do evento a qual as odds se referem

    /**
     * CONSTRUTORES
     */
    public Apostador(String email, String password, String nome, double valor) {
        super(email, password, nome);
        this.contadorIdAposta = 0;
        this.valor = valor;
        this.listaApostas = new TreeMap<>();
        this.notificacoesOdd = new TreeSet<>();
    }

    /**
     * GETTERS E SETTERS
     */
    public double getValor() {
        return valor;
    }

    public void setValor(double valor) {
        this.valor = valor;
    }

    public void addContadorAposta() {
        this.contadorIdAposta++;
    }

    public int getContadorAposta() {
        return this.contadorIdAposta;

    }

    /**
     * Adicionar Aposta
     */
    public void addAposta(int aposta, Aposta apos) {
        // System.out.println("Odd Aposta Apotador:" + apos.getOdd().getOddB());
        this.listaApostas.put(apos.getId(), apos);
        this.contadorIdAposta++;
    }

    public boolean containsEvent(int id) {
        boolean flag = false;
        for (int idAposta : this.listaApostas.keySet()) {
            if (this.listaApostas.get(id).getIdEvento() == id) {
                flag = true;
                break;
            }

        }
        return flag;
    }

    public ArrayList<Aposta> getApostasByIdEvent(int evento) {
        ArrayList<Aposta> aux = new ArrayList<>();
        for (int idAposta : this.listaApostas.keySet()) {
            if (this.listaApostas.get(idAposta).getIdEvento() == evento) {
                aux.add(this.listaApostas.get(idAposta));
            }
        }
        return aux;
    }

    /**
     * Obter Valor Ganho na Aposta
     */
    public double getGanhoAposta(Aposta apos) {
        return this.listaApostas.get(apos.getId()).getGanho();
    }

    /**
     * Obter Lista Apostas
     *
     * @return
     */
    public TreeMap<Integer, Aposta> getListaApostas() {
        return this.listaApostas;
    }

    public void setResultadoAposta(int idAposta, int resultadoAposta) {
        this.listaApostas.get(idAposta).setResultado(resultadoAposta);
        this.listaApostas.get(idAposta).setGanho();
        //this.listaApostas.get(idAposta).setEstadoOff();
    }

    public String ganhoApostasToString() {
        StringBuilder s = new StringBuilder();
        for (int id : this.listaApostas.keySet()) {
            Aposta a = this.listaApostas.get(id);
            /*
             if (a.getEstado()) {
             s.append(a.toString()).append("\n");
             s.append("Ganho = ").append(a.getGanho());

             }
             */

        }
        return s.toString();

    }

    public TreeSet<Integer> getNotificacoesOdd() {
        return this.notificacoesOdd;
    }

    public void setNotificacaoOdd(int idEvento) {
        this.notificacoesOdd.add(idEvento);

    }

    public void limpaNoficacoesOdd() {
        this.notificacoesOdd = new TreeSet<>();
    }

    @Override
    public void update(int idEvento, String evento,int resultado) {
        for (Aposta apos : this.getApostasByIdEvent(idEvento)) {
            apos.setFechada();
            apos.setResultado(resultado);
            apos.calGanho();
            this.valor += apos.getGanho();
        }
    }

    @Override
    public void updateOdd(int idEvento) {
        this.notificacoesOdd.add(idEvento);

    }

}

File: Ruby_Observer/apostador.rb
require_relative 'user'
require_relative 'aposta'
require_relative 'evento'
require 'time'
require_relative 'self_observer'

require 'set'

class Apostador < User
  include SelfObserver
  #implements :SelfObserver

  attr_accessor :nome, :valor, :not_odd, :lista_apostas, :contador_id_aposta

  def initialize(email, password, nome, valor)
    super(email, password, nome)
    @valor=valor
    @contador_id_aposta = 0
    @lista_apostas = Hash.new
    @not_odd = Set.new
  end

  def add_contador_aposta
    @contador_id_aposta += 1
  end

  #def add_aposta(id_aposta,aposta)
  def add_aposta(aposta)
    @lista_apostas[aposta.get_id] = aposta
    add_contador_aposta
  end

  def contains_event(id)
    flag = false
    @lista_apostas.keys.each do |id_aposta|
      flag = true if id_aposta.eql?(id)
    end
    return flag
  end

  def get_apostas_by_id_event(evento)
    apostas = Set.new
    @lista_apostas.keys.each do |id_aposta|
      apostas.add(@lista_apostas[id_aposta]) if id_aposta.eql?(evento)
    end
    return apostas
  end

  def set_resultado_aposta(id_aposta, resultado_aposta)
    @lista_apostas[id_aposta].set_resultado(resultado_aposta)
    @lista_apostas[id_aposta].cal_ganho
  end

  def limpa_noficacoes_odd
    @not_odd = Set.new
  end

  def update(id, info, resultado)

    self.get_apostas_by_id_event(id.to_i).each do |apos|
      apos.set_fechada
      apos.set_resultado(resultado)
      apos.cal_ganho

      @valor += apos.ganho.to_f

    end
  end


  def update_odd(id_evento)
    @not_odd.add(id_evento)
  end

  def set_valor(valor)
    @valor = valor
  end
end
